rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/SpatialMatcher.scala

```diff
@@ -264,119 +248,130 @@ object SpatialMatcher extends SpatialMatcherInstances {
+    //Correctness of the matching relies on them being sorted.
+    //Use a type class? A type tag?
+    val termsSorted    = tlist
+    val patternsSorted = plist.zip(Stream.continually(termsSorted))
+
+    def guard(predicate: => Boolean): OptionalFreeMapWithCost[Unit] =
+      if (predicate) OptionalFreeMapWithCost.pure(()) else OptionalFreeMapWithCost.emptyMap
+
+    val maximumBipartiteMatch = MaximumBipartiteMatch((p: T, t: T) => {
+      val matchEffect = if (!lf.connectiveUsed(p)) {
+        guard(t == p).modifyCost(_.charge(COMPARISON_COST))
+      } else {
+        spatialMatch(t, p)
       }
-    }
+      matchEffect.attempt.map(_.isRight)
+    })
+
+    for {
+      matchedTargetsOpt <- maximumBipartiteMatch.findMatches(patternsSorted.toList)
+      matchedTargets <- matchedTargetsOpt.fold(
+                         OptionalFreeMapWithCost.emptyMap[Seq[T]]
+                       )(
+                         OptionalFreeMapWithCost.pure[Seq[T]]
+                       )
+      remainderTargets = termsSorted.filterNot(matchedTargets.contains)
+      _ <- varLevel match {
+            case None =>
+              // If there is a wildcard, we succeed.
+              // We also succeed if there isn't but the remainder is empty.
+              if (wildcard || remainderTargets.isEmpty)
+                OptionalFreeMapWithCost.pure(())
+              else
+                // This should be prevented by the length checks.
+                OptionalFreeMapWithCost.emptyMap
+            // If there's a capture variable, we prefer to add things to that rather than throw them away.
+            case Some(level) => {
+              /*
+              Matching a single variable with a single term has so far been handled
+              by matching a list of 0 patterns with a list of 1 target, with a remainder present.
+
+              In general, matching remainders requires combining/merging all terms matched to the remainder
+              on a single capture variable. E.g. `1 | 2 | 3 matches 2 | x` would return x = 1 | 3.
+              See the `handleReminder` method below and notice that initial state for the fold is taken from
+              previous matches.
+
+              Matching a single variable using that mechanism (e.g. [1] matches [x]) was OK so far, since
+              overriding the match state (FreeMap entry) for a given variable never happened in the previous algo.
+
+              We don't have a proof it was impossible. What we know for sure though is:
+              none of the tests added for the new algo - which revealed this problem - have failed on the old algo.
+
+              Since we've started modifying the var assignment during the MBM algorithm, the case of matching a var
+              to a single term must be handled separately, without merging the prior match state.
+              */
+              if (patternsSorted.isEmpty && remainderTargets.length == 1 &&
```

Comment:
@guardbotmk3 This needs your special attention and feedback. I'm not entirely sure if the condition used here is right or why the issue wouldn't surface earlier.

I tried getting rid of the `noFrees` in expressions matching and match the free variables in the expression matcher, but then thought we should discuss if it's the right approach.

Reply:
> Since we've started modifying the var assignment during the MBM algorithm, the case of matching a var
> to a single term must be handled separately, without merging the prior match state.

You just explained why this is wrong:
> we've started modifying the var assignment during the MBM algorithm

This **should not** be the case. For each (target, pattern) pair you test, you should find the update that results when they are matched, and you should apply the updates from **only** those (target, pattern) pairs in the final match